qubesadmin/tests/tools/qvm_run.py

```diff
@@ -643,3 +643,179 @@ def test_021_paused_vm(self):
+            b'0\x00power_state=Running'
+        ret = qubesadmin.tools.qvm_run.main(
+            ['--no-gui', '--no-shell', 'test-vm', 'command', '--', 'arg'],
+            app=self.app)
+        self.assertEqual(ret, 0)
+        self.assertEqual(self.app.service_calls, [
+            ('test-vm', 'qubes.VMExec+command+arg', {
+                'stdout': subprocess.DEVNULL,
+                'stderr': subprocess.DEVNULL,
+                'user': None,
+            }),
+            ('test-vm', 'qubes.VMExec+command+arg', b'')
+        ])
+        self.assertAllCalled()
+
+    def test_025_no_shell_double_dashdash(self):
+        self.app.expected_calls[
+            ('dom0', 'admin.vm.List', None, None)] = \
+            b'0\x00test-vm class=AppVM state=Paused\n'
+        self.app.expected_calls[
+            ('test-vm', 'admin.vm.feature.CheckWithTemplate',
+             'vmexec', None)] = \
+            b'0\x001'
+        self.app.expected_calls[
+            ('test-vm', 'admin.vm.CurrentState', None, None)] = \
+            b'0\x00power_state=Running'
+        ret = qubesadmin.tools.qvm_run.main(
+            ['--no-gui', '--no-shell', 'test-vm', 'command', '--', '--', 'arg'],
+            app=self.app)
+        self.assertEqual(ret, 0)
+        self.assertEqual(self.app.service_calls, [
+            ('test-vm', 'qubes.VMExec+command+----+arg', {
+                'stdout': subprocess.DEVNULL,
+                'stderr': subprocess.DEVNULL,
+                'user': None,
+            }),
+            ('test-vm', 'qubes.VMExec+command+----+arg', b'')
+        ])
+        self.assertAllCalled()
+
+    def test_026_no_shell_double_dashdash(self):
+        self.app.expected_calls[
+            ('dom0', 'admin.vm.List', None, None)] = \
+            b'0\x00test-vm class=AppVM state=Paused\n'
+        self.app.expected_calls[
+            ('test-vm', 'admin.vm.feature.CheckWithTemplate',
+             'vmexec', None)] = \
+            b'0\x001'
+        self.app.expected_calls[
+            ('test-vm', 'admin.vm.CurrentState', None, None)] = \
+            b'0\x00power_state=Running'
+        ret = qubesadmin.tools.qvm_run.main(
+            ['--no-gui', '--no-shell', '--', 'test-vm', 'command', '--', 'arg'],
+            app=self.app)
+        self.assertEqual(ret, 0)
+        self.assertEqual(self.app.service_calls, [
+            ('test-vm', 'qubes.VMExec+command+----+arg', {
+                'stdout': subprocess.DEVNULL,
+                'stderr': subprocess.DEVNULL,
+                'user': None,
+            }),
+            ('test-vm', 'qubes.VMExec+command+----+arg', b'')
+        ])
+        self.assertAllCalled()
+
+    def test_027_no_shell_dispvm(self):
+        self.app.expected_calls[
+            ('$dispvm:test-vm', 'admin.vm.feature.CheckWithTemplate',
+             'vmexec', None)] = \
+            b'0\x001'
+        ret = qubesadmin.tools.qvm_run.main(
+            ['--no-gui', '--dispvm', 'test-vm', 'command', '--', 'arg'],
```

Comment:
Written this way, `test-vm` should be really interpreted as a command (and no explicit base VM for DispVM - use default). If you want explicit base VM, then use `--dispvm=test-vm`. I'm not sure if argparse can make this distinction...